### PR TITLE
Tests and fixes

### PR DESCRIPTION
**Pull request by mattiekat**
Wrote in some new tests and fixed a misunderstanding I had that lead to some issues

Fixes abacus-network/abacus-monorepo#1026

### CODE REVIEW

File: ethers-providers/src/transports/quorum.rs [hunk 1]
```diff
@@ -685,6 +682,95 @@ mod tests {
...
     }
 
+    #[tokio::test]
+    async fn test_get_quorum_block_number() {
+        let mut providers = Vec::new();
+        let mut mocked = Vec::new();
+
+        for value in [100, 101, 68, 100, 102] {
+            let mock = MockProvider::new();
+            for _ in 0..6 {
```
Comment (tkporter) [inline ethers-providers/src/transports/quorum.rs:691]: this 6 because there are 6 expected RPCs to each provider (i.e. 6 calls to `get_quorum_block_number` in this test) right?

File: ethers-providers/src/transports/quorum.rs [hunk 2]
```diff
@@ -685,6 +682,95 @@ mod tests {
...
+                mock.push(U64::from(value)).unwrap();
+            }
+            providers.push(WeightedProvider::new(mock.clone()));
+            mocked.push(mock);
```
Comment (tkporter) [inline ethers-providers/src/transports/quorum.rs]: Does this need to be here?

File: ethers-providers/src/transports/quorum.rs [hunk 3]
```diff
@@ -685,6 +682,95 @@ mod tests {
...
+            .add_providers(providers.clone())
+            .quorum(Quorum::Majority)
+            .build();
+        assert_eq!(quorum.get_quorum_block_number().await.unwrap().as_u64(), 100);
+    }
+
+    #[tokio::test]
+    async fn test_get_quorum_block_number_with_errors() {
+        let mut providers = Vec::new();
+        let mut mocked = Vec::new();
```
Comment (tkporter) [inline ethers-providers/src/transports/quorum.rs]: ditto I don't see this being used other than being pushed to?

File: ethers-providers/src/transports/quorum.rs [hunk 4]
```diff
@@ -685,6 +682,95 @@ mod tests {
...
+        }
+        providers.push(WeightedProvider::new(mock.clone()));
+        mocked.push(mock);
+
+        let quorum = QuorumProvider::builder()
+            .add_providers(providers.clone())
+            .quorum(Quorum::ProviderCount(2))
+            .build();
+        assert_eq!(quorum.get_quorum_block_number().await.unwrap().as_u64(), 100);
+
```
Comment (tkporter) [inline ethers-providers/src/transports/quorum.rs:760]: While we're at it I think we should add a test case where `quorum.get_quorum_block_number()` doesn't find a quorum

File: ethers-providers/src/transports/quorum.rs [hunk 5]
```diff
@@ -277,23 +277,20 @@ pub enum Quorum {
...
 
 impl Quorum {
-    fn weight<'a, T: 'a>(
-        self,
-        providers: impl IntoIterator<Item = &'a WeightedProvider<T>>,
-    ) -> u64 {
+    fn weight<T>(self, providers: &[WeightedProvider<T>]) -> u64 {
         match self {
-            Quorum::All => providers.into_iter().map(|p| p.weight).sum::<u64>(),
+            Quorum::All => providers.iter().map(|p| p.weight).sum::<u64>(),
```
Comment (tkporter) [inline ethers-providers/src/transports/quorum.rs:282]: iirc this was previously `iter()` and then a previous PR changed to `into_iter()` - wondering why the change back?

**Review by tkporter — APPROVED** [review]
lgtm but just some small nits - could you also share what the issue was? It's not immediately clear to me how the semantics have changed to fix things

**Comment by tkporter** [comment]
btw I think this may auto-close https://github.com/abacus-network/abacus-monorepo/issues/1026 so just flagging that we should make sure to do the other bits mentioned in there too

File: ethers-providers/src/transports/quorum.rs (hunk 1 above)
Comment (mattiekat) [reply ethers-providers/src/transports/quorum.rs:691]: Yes, you push each time you want it to have a response, otherwise it gives a no response error as part of the mock implementation. (I do use this fact in one of the tests).

File: ethers-providers/src/transports/quorum.rs (hunk 2 above)
Comment (mattiekat) [reply ethers-providers/src/transports/quorum.rs]: Cleaned up

File: ethers-providers/src/transports/quorum.rs (hunk 5 above)
Comment (mattiekat) [reply ethers-providers/src/transports/quorum.rs:282]: Because I undid changes in an earlier commit and this was how they had it written.

File: ethers-providers/src/transports/quorum.rs (hunk 4 above)
Comment (mattiekat) [reply ethers-providers/src/transports/quorum.rs:760]: Done
